### ISSUE TITLE
Topology validation/normalization using Schema

### ISSUE THREAD

**Issue by jbemmel** [ipspace/netlab#521]
[Schema](https://pypi.org/project/schema/) could be a useful tool to help validate and normalize the topology input, before handing it over to the various modules for processing and validation.

Instead of writing custom pre_validate hooks per module, having the vrf module check that vrf names are at most 16 characters, etc. just pass Schema over the topology in an overall validation/normalization step. Replace empty vrfs with ```{}```, check syntax, referential integrity, etc.

It's still Python code, not some YANG model exercise (would advise against that), but it helps to separate out some bugs/issues/concerns that are currently mixed in with the transformation code

**Comment by ipspace** [comment]
Schema looks like the least awful validation tool I've seen so far, would be interesting to see how useful its error messages are -- the usual error messages I get from validation tools are pretty much useless the moment the schema gets a bit more complex.

If we figure out it fits the bill, it wouldn't make sense to do things two different ways in different parts of the code. Would you be willing to rewrite all validation logic to use Schema and create the necessary validation tests?

**Comment by jbemmel** [comment]
I can start a feature branch and experiment with this, to see if we like it. Am thinking to put this in front i.e. in CLI read.py
```
def run(cli_args: typing.List[str]) -> None:
  args = read_topology_parse(cli_args)
  common.set_logging_flags(args)
  topology = read_topology.load(args.topology,args.defaults,"package:topology-defaults.yml")

  if 'settings' in args:
    topology.nodes = augment.nodes.create_node_dict(topology.nodes)
    read_topology.add_cli_args(topology,args)

  common.exit_on_error()

  validated_topology = validation.validate( topology )
  if validated_topology:
    transform_setup(validated_topology)
    args.output.write(validated_topology.to_yaml())
```

and then gradually build out the types of checks performed in ```validate```, removing them from ```transform_setup```.

Starting with missing attribute validation checks, everything based on topology-defaults of course ( not redefine the structure of the schema )

**Comment by ipspace** [comment]
> I can start a feature branch and experiment with this, to see if we like it. Am thinking to put this in front i.e. in CLI read.py

One of the unexplored use cases of **netsim** modules is to use them as a library. Data validation should therefore happen at the beginning of the transformation process.

> and then gradually build out the types of checks performed in `validate`, removing them from `transform_setup`.

In the ideal world, each module would contribute its own subschema for global, node, link, and interface attributes.

> Starting with missing attribute validation checks, everything based on topology-defaults of course ( not redefine the structure of the schema )

I don't think we'd be able to retain attribute lists in topology-defaults while having useful schemata that do more than just keyword validation (or we'd have duplication of information). That's why I'm saying we should be doing one or the other, but not a bit of both.

In any case, a simple PoC in a validate hook (once I add that) would be more than good enough to see how easy it would be to use the tool and how helpful the error messages are.

Having written a programming language in my distant past, I'm a bit obsessed with easy-to-interpret error messages that point to the source of the problem instead of saying "you have a problem".

**Comment by ipspace** [comment]
Similar solution implemented in #713
